File: checkio_js/home/javascript/frequencySort.js
import assert from "assert";

function frequencySort(items) {
    let set_arr = []
    items.forEach(a => set_arr.includes(a) ? 0 : set_arr.push(a))
    let arr = []
    set_arr.forEach(a => arr.includes(a) ? 0 : arr.push(Array(items.filter(b => b == a).length).fill(a)))
    arr.sort((a, b) => b.length - a.length)
    let final_arr = []
    arr.forEach(a => final_arr.push(...a))
    return final_arr
    /**
    let arr = []
    items.forEach(a => arr.includes(a)? 0 : arr.push(a))
    let sorted = []
    arr.forEach(a => sorted.push(...items.filter(b => b === a)))
    return sorted
    **/
}

console.log('Example:');
console.log(frequencySort([4, 6, 2, 2, 6, 4, 4, 4]));

// These "asserts" are used for self-checking and not for an auto-testing
assert.deepEqual(frequencySort([4, 6, 2, 2, 6, 4, 4, 4]), [4, 4, 4, 4, 6, 6, 2, 2]);
assert.deepEqual(frequencySort(['bob', 'bob', 'carl', 'alex', 'bob']), ['bob', 'bob', 'bob', 'carl', 'alex']);
assert.deepEqual(frequencySort([17, 99, 42]), [17, 99, 42]);
assert.deepEqual(frequencySort([]), []);
assert.deepEqual(frequencySort([1]), [1]);
assert.deepEqual(frequencySort([4,6,2,2,2,6,4,4,4]), [4,4,4,4,2,2,2,6,6]);

console.log("Coding complete? Click 'Check' to earn cool rewards!");

File: checkio_js/home/python/frequencySort.py


def frequencySort(items):
    d = {}
    for i in items:
        d[i] = 1 if i not in d else d[i] + 1

    '''
    for i in items:
        if i not in d:
            d[i] = 1
        else:
            d[i] += 1
    '''

    '''
    for i in items:
        if i not in d:
            d[i] = items.count(i)
    '''

    dsort = sorted(d.items(), key=lambda x: x[1], reverse=True)
    return [a for x in [[y]*z for y, z in dsort] for a in x]


if __name__ =='__main__':
    print('Example:');
    print(frequencySort([4, 6, 2, 2, 6, 4, 4, 4]));

# These "asserts" are used for self-checking and not for an auto-testing
assert frequencySort([4, 6, 2, 2, 6, 4, 4, 4]) == [4, 4, 4, 4, 6, 6, 2, 2]
assert frequencySort(['bob', 'bob', 'carl', 'alex', 'bob']) == ['bob', 'bob', 'bob', 'carl', 'alex']
assert frequencySort([17, 99, 42]) == [17, 99, 42]
assert frequencySort([]) == []
assert frequencySort([1]) == [1]
assert frequencySort([4,6,2,2,2,6,4,4,4]) == [4,4,4,4,2,2,2,6,6]

print("Coding complete? Click 'Check' to earn cool rewards!");
